qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,529 @@
+    def _js_slot(*args):
+        """Wrap a methos as a JavaScript function.
+
+        Register a PACContext method as a JavaScript function, and catch
+        exceptions returning them as JavaScript Error objects.
+
+        Args:
+            args: Types of method arguments.
+
+        Return: Wrapped method.
+        """
+        def _decorator(method):
+            @functools.wraps(method)
+            def new_method(self, *args, **kwargs):
+                try:
+                    return method(self, *args, **kwargs)
+                except:
+                    traceback.print_exc()
+                    e = sys.exc_info()[0]
+                    #pylint: disable=protected-access
+                    return self._error_con.callAsConstructor([str(e)])
+            return pyqtSlot(*args, result=QJSValue)(new_method)
+        return _decorator
+
+    def __init__(self, engine, fixed_dt=None):
+        """Create a new PAC API implementation instance.
+
+        Args:
+            engine: QJSEngine which is used for running PAC.
+            fixed_dt: Assume fixed date and time instead of system time.
+        """
+        QObject.__init__(self)
+        self._engine = engine
+        self._error_con = engine.globalObject().property("Error")
+        self._fixed_dt = fixed_dt
+
+    @_js_slot(str)
+    def isPlainHostName(self, host):
+        """Test is host is a plain hostname.
+
+        Return True if and only if there is no domain name in the hostname (no
+        dots).
+
+        Args:
+            host: The hostname from the URL (excluding port number).
+        """
+        return '.' not in host
+
+    @_js_slot(str, str)
+    def dnsDomainIs(self, host, domain):
+        """Test is host belongs to domain.
+
+        Return true if and only if the domain of hostname matches.
+
+        Args:
+            host: The hostname from the URL.
+            domain: The domain name to test the hostname against.
+        """
+        return host.endswith(domain)
+
+    @_js_slot(str, str)
+    def localHostOrDomainIs(self, host, hostdom):
+        """Test if host is local or belongs to domain.
+
+        Return true if the hostname matches exactly the specified hostname, or
+        if there is no domain name part in the hostname, but the unqualified
+        hostname matches.
+
+        Args:
+            host: The hostname from the URL.
+            hostdom: Fully qualified hostname to match against.
+        """
+        (_, dot_host, _) = host.partition('.')
+        (unqual_hostdom, _, _) = hostdom.partition('.')
+        if dot_host == "":
+            return host == unqual_hostdom
+        else:
+            return host == hostdom
+
+    @_js_slot(str)
+    def isResolvable(self, host):
+        """Test if hostname is resolvable.
+
+        Tries to resolve the hostname. Return true if succeeds.
+
+        Args:
+            host: The hostname from the URL.
+        """
+        return self.dnsResolve(host) != ""
```

Comment:
Hm, personally I like explicit checks against values of the same type because that way it's more readable for me (it conveys more type information). If you prefer this way though I'll change.